### DISCUSSION TITLE
Underscore for custom methods in feathers^5

### DISCUSSION THREAD

**Discussion by fratzinger** [feathersjs/feathers#2304] [Ideas]
Hey :)

Will custom methods from feathers^5 also support underscore-functions which will skip the hooks entirely? Like the methods in adapter-commons: `_find` that gets called automatically by `service('posts').find()` call? As I start to integrate the custom methods to my libraries in preparation of v5 I wondered if I should start all my methods with underscore. For example for `feathers-authentication-management`: `_checkUnique(...)` etc.

The ability to skip hooks with `_find`, `_get`, ... comes in pretty handy sometimes. so that would be nice to have for custom methods right out of the box.

What's your opinion on that, David? Thanks in advance for your thoughts! As always I cannot tell you enough, how I appreciate the effort you put into making this framework!

**Comment by daffl** [comment]
Are those methods that should be accessed externally? For custom methods you can basically use any name other than the standard service methods but the signature will always be `(data, params)`. A custom method will always be enabled for hooks (so e.g. application level hooks will still run) but only have their own. For authentication management specifically I'm thinking things like `resendVerification` or maybe even `resetPassword`.

**Reply by fratzinger** [reply]
Thanks for the quick answer! 
Yes, these methods will be exposed externally. `(data, params)` fits quite well, because `feathers-authentication-management` as of now only uses `create`.
I would like to enable all existing "actions" to be exposed as custom methods.

The possibility to call `_find` etc. only exists on the server side, of course. It would be nice to have a similar behavior for custom methods on the server side: that you call them without any hooks at all (for example `_resendVerification`).

**Reply by daffl** [reply]
Do you expect them to have a lot of external hooks (other than maybe authentication)? What the service ones do is basically this:

```js
class SomeService {
  _myMethod(data, params) {}

  myMethod (data, params) {
    return this._myMethod(data, params);
  }
}
```

**Reply by fratzinger** [reply]
Maybe logging, permissions. My point is not specifically about auth-management but to have a standardized way / best practice that most developers follow. So for every library I know I can use the underscore functions (as for every Service that extends "AdapterService").

Maybe even a more dynamic way to think about would be to have a `skipHooks` prop in `params` like it is in "feathers-fetching"

**Reply by tvld** [reply]
To weigh-in for a moment, I am not sure if usage of "underscore" would make things more clear. As for myself it would seem an anti-pattern for feathers.. but that is maybe because I do not know what it is and never had need for it ... maybe ))
